Make the `build-catalogue` binary work with C++ source files.

## Goal
Let end users directly write their mechanisms against the mech-ABI and compile it with the build-catalog script supplied by arbor. 

## Rationale
It will be an overall easier way of writing custom C++ mechanisms. Which doesn't has the need to write the Parent catalog file by the end user. Making it more portable. Also give a great way to select the mechanisms that the user want's to compile in trough the CMD arguments. 

## Scope
change: 
- Build-catalogue
- BuildModules.cmake

## Implementation

# build-catalogue (diff):

I would suggest to not use autodiscovery as done for nmod but list the mechanisms that needs to be compiled in.

```
parser.add_argument('-c', '--CXX',
                    nargs="*",
                    help='pass list of cxx mechs')
~.....

if(args['CXX']):
    mods    = args['CXX']  # list of the mechanisms get loaded here.

~.....

make_catalogue(
  NAME {name}
  SOURCES "${{CMAKE_CURRENT_SOURCE_DIR}}/mod"
  OUTPUT "CAT_{name.upper()}_SOURCES"
  MECHS {' '.join(mods)}
  CXXIN {"TRUE" if CXX else "FALSE"}
  PREFIX /usr/local/share/arbor
  CXX_FLAGS_TARGET ${{ARB_CXX_FLAGS_TARGET}}
  STANDALONE ON
  VERBOSE {"ON" if verbose else "OFF"})    
```

# BuildModules.cmake (diff)

- add extra argument. eg. CXXIN  (boolean)

solve it like this:
```
  if(MK_CAT_CXXIN)
    set(MK_CAT_OUT_DIR ${MK_CAT_SOURCES})
    add_custom_target(build_catalogue_${MK_CAT_NAME}_mods)
  else()
    file(MAKE_DIRECTORY "${MK_CAT_OUT_DIR}")
    build_modules(
      ${MK_CAT_MECHS}
      SOURCE_DIR "${MK_CAT_SOURCES}"
      DEST_DIR "${MK_CAT_OUT_DIR}"
      ${external_modcc} # NB: expands to 'MODCC <binary>' to add an optional argument
      MODCC_FLAGS -t cpu -t gpu ${ARB_MODCC_FLAGS} -N arb::${MK_CAT_NAME}_catalogue
      GENERATES .hpp _cpu.cpp _gpu.cpp _gpu.cu
      TARGET build_catalogue_${MK_CAT_NAME}_mods)
  endif()
  ```
  
 Usage would look like:
 ```
 make_catalogue(
        NAME EMC
        SOURCES "${CMAKE_CURRENT_SOURCE_DIR}/EMC/EMC_Mechs"  # redirected to output path ++ the emc_cat.cpp
        OUTPUT "CAT_EMC_SOURCES"                             #    see above -
        CXX_FLAGS_TARGET "${ARB_CXX_FLAGS_TARGET_FULL}"
        STANDALONE FALSE
        VERBOSE TRUE
        CXXIN   TRUE
        MECHS ou_noise gj_linear adex adex_DCN adex_PC calshadow
        PREFIX "${PROJECT_SOURCE_DIR}/mechanisms"            #   Unused NOW
)
```

Just now I have also tried to employ a raw mechanism along with NMODL mechanisms using `arbor-build-catalogue` from v.0.6.0. As a first simple test, I used for the raw mechanism the `.cpp`/`.hpp`/`.cu` files that I got from running `arbor-build-catalogue` in debug mode on `.mod` files only. Then, after removing `expsyn_curr_calcium_plasticity.mod` and adding the raw `.cpp`/`.hpp`/`.cu` files, my `mechanisms` directory contains the following:
```
deltasyn.mod                            expsyn_curr_calcium_plasticity_gpu.cpp  expsyn_curr_calcium_plasticity.hpp  lif.mod
expsyn_curr_calcium_plasticity_cpu.cpp  expsyn_curr_calcium_plasticity_gpu.cu   expsyn_curr.mod                     ornstein_uhlenbeck.mod
```
However,  running `arbor-build-catalogue custom mechanisms --raw "expsyn_curr_calcium_plasticity"` gives me the following error:
```
Building catalogue 'custom' from mechanisms in /home/jlubo/Desktop/Arbor_Network_GitLab_Repository/mechanisms
 * NMODL
   * expsyn_curr
   * lif
   * deltasyn
   * ornstein_uhlenbeck
 * Raw
   * expsyn_curr_calcium_plasticity
Build log:
-- The CXX compiler identification is GNU 10.3.0
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Configuring done
Traceback (most recent call last):
  File "/home/jlubo/local_dev/bin/arbor-build-catalogue", line 231, in <module>
    sp.run(cmake_cmd, shell=True, check=True, stdout=out, stderr=err)
  File "/usr/lib/python3.8/subprocess.py", line 516, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command 'cmake ..' returned non-zero exit status 1.
 Error:

CMake Error at BuildModules.cmake:129 (add_library):
  Cannot find source file:

    /tmp/tmpr7_52yai/build/generated/custom/expsyn_curr_calcium_plasticity_cpu.cpp

  Tried extensions .c .C .c++ .cc .cpp .cxx .cu .mpp .m .M .mm .ixx .cppm .h
  .hh .h++ .hm .hpp .hxx .in .txx .f .F .for .f77 .f90 .f95 .f03 .hip .ispc
Call Stack (most recent call first):
  CMakeLists.txt:19 (make_catalogue)


CMake Error at BuildModules.cmake:129 (add_library):
  No SOURCES given to target: custom-catalogue
Call Stack (most recent call first):
  CMakeLists.txt:19 (make_catalogue)


CMake Generate step failed.  Build files cannot be regenerated correctly.
```
Using debug mode, I find that while the file `expsyn_curr_calcium_plasticity.hpp` somehow makes it into the target directory `tmp*/build/generated/custom/`, the `.cpp` and `.cu` files only make it to the `tmp*/mod/` directory.

--- a/scripts/build-catalogue.in
+++ b/scripts/build-catalogue.in
@@ -67,8 +67,8 @@ also <name>_gpu.cpp and <name>_gpu.cu (if not given -C).''')
                         help='Enable GPU support, valid options: cuda|hip|cuda-clang.')
 
     parser.add_argument('-C', '--no-cpu',
-                        action='store_false',
-                        default=True,
+                        action='store_true',
+                        default=False,
                         help='Disable CPU support.')
 
     parser.add_argument('-d', '--debug',

--- a/scripts/build-catalogue.in
+++ b/scripts/build-catalogue.in
@@ -68,7 +68,6 @@ also <name>_gpu.cpp and <name>_gpu.cu (if not given -C).''')
 
     parser.add_argument('-C', '--no-cpu',
                         action='store_true',
-                        default=False,
                         help='Disable CPU support.')
 
     parser.add_argument('-d', '--debug',

